skunk/rust/skunk/src/parser/tests/expression_builder.rs

```diff
@@ -0,0 +1,70 @@
+
+use super::super::ast::{ExpressionValue, ExpressionValueEnum, Operator, OutputExpression, SafeSpan};
+
+pub struct Expr<'a> {
+  offset: usize,
+  line: u32,
+  expr: Box<dyn 'a + FnOnce(usize, u32) -> ExpressionValueEnum>,
+}
+
+impl <'a> Expr<'a> {
+  fn mk(self, offset: usize, line: u32) -> ExpressionValue {
+    let offset = offset + self.offset;
+    let line = line + self.line;
+    ExpressionValue { 
+      position: SafeSpan { offset, line },
+      info: (self.expr)(offset, line),
+    }
+  }
+  pub fn block(offset: usize, line: u32, mut body: Vec<Expr<'a>>) -> Self {
+    Expr { expr: Box::new(move |offset, line| ExpressionValueEnum::Block(body.drain(..).map(|e| e.mk(offset, line)).collect())), offset, line }
+  }
+  pub fn output(offset: usize, line: u32, name: &'a str, expr: Expr<'a>) -> Self {
+    Expr { expr: Box::new(move |offset, line| ExpressionValueEnum::Output(OutputExpression { output: name.to_string(), expression: Box::new(expr.mk(offset, line)), and_return: false })), offset, line }
+  }
+  pub fn sref(offset: usize, line: u32, name: &'a str) -> Self {
+    Expr { expr: Box::new(move |_, _| ExpressionValueEnum::ReferenceToState(name.to_string())), offset, line }
+  }
+  pub fn empty(offset: usize, line: u32) -> Self {
+    Expr { expr: Box::new(move |_, _| ExpressionValueEnum::Empty), offset, line }
+  }
+  pub fn fun(offset: usize, line: u32, name: &'a str, expr: Expr<'a>) -> Self {
+    Expr { expr: Box::new(move |offset, line| ExpressionValueEnum::FunctionCall(name.to_string(), Box::new(expr.mk(offset, line)))), offset, line }
+  }
+  pub fn op(self, offset: usize, line: u32, op: Operator, rhs: Expr<'a>) -> Self {
+    let offset = offset + self.offset;
+    let line = line + self.line;
+    Expr { expr: Box::new(move |prev_offset, prev_line| {
+      let lhs = self.mk(prev_offset - offset, prev_line - line);
+      let rhs = rhs.mk(prev_offset, prev_line);
+      ExpressionValueEnum::BinaryOperator(Box::new(lhs), op, Box::new(rhs))
+    }), offset, line }
+  }
+  pub fn tuple_ref(self, offset: usize, line: u32, idx: i64) -> Self {
+    let offset = offset + self.offset;
```

Comment:
Suggestion for future:

Might be worth breaking the line & offset pairs out into a separate struct so you can have a slightly more readable:
`let new_location = self.location.advance(location)`

Reply:
added a TODO.